### PR TITLE
LNX-147 Add tags for boolean values

### PR DESCRIPTION
**Pull request by EasyJB**
Tags are necessary if we don't want to change object class in 2 modules each time we add new feature.  
Map generator still needs to be adapted to this logic.

### CODE REVIEW

File: lynx/common/actions/chop.py [hunk 1]
```diff
@@ -23,19 +23,18 @@ def apply(self, scene: 'Scene') -> None:
         objects_on_square = scene.get_objects_by_position(target_position)
         # TODO put 'Tree' into an Enum
         chopped_object: Object = \
-        list(filter(lambda object_on_square: object_on_square.name == 'Tree', objects_on_square))[0]
+            list(filter(lambda object_on_square: object_on_square.name == 'Tree', objects_on_square))[0]
```
Comment (filip-szweda) [inline lynx/common/actions/chop.py]: obraz

Comment (filip-szweda) [inline lynx/common/actions/chop.py]: Make in-line with `chopped_object: Object =` i.e.
`chopped_object: Object = list(filter(lambda object_on_square: object_on_square.name == 'Tree', objects_on_square))[0]`
Same for other lines that fit in 150 signs.

File: lynx/common/actions/chop.py [hunk 2]
```diff
@@ -23,19 +23,18 @@ def apply(self, scene: 'Scene') -> None:
...
         # TODO put 'Tree' into an Enum
         chopped_object: Object = \
-        list(filter(lambda object_on_square: object_on_square.name == 'Tree', objects_on_square))[0]
+            list(filter(lambda object_on_square: object_on_square.name == 'Tree', objects_on_square))[0]
         if chopped_object:
             remove_action = RemoveObject(chopped_object.id)
-            log = Object(id=scene.generate_id(), name='Log', pushable=True, walkable=False,
-                         position=chopped_object.position,
-                         pickable=True)
+            log = Object(id=scene.generate_id(), name='Log', tags=['pushable', 'walkable', 'pickable'],
```
Comment (filip-szweda) [inline lynx/common/actions/chop.py]: Not `walkable`, see `walkable=False`

File: lynx/common/actions/common_requirements.py [hunk 3]
```diff
@@ -2,45 +2,57 @@
 from lynx.common.square import Square
 from lynx.common.vector import Vector
 from lynx.common.enums import Direction
+from typing import List
```
Comment (filip-szweda) [inline lynx/common/actions/common_requirements.py]:
```
from typing import List

from lynx.common.object import Object
from lynx.common.square import Square
from lynx.common.vector import Vector
from lynx.common.enums import Direction
```

File: lynx/common/actions/common_requirements.py [hunk 4]
```diff
@@ -2,45 +2,57 @@
...
+    # TODO: Move, Push, Chop etc. should take argument of Direction type, not Vector type
+    # this means that serialization and deserialization have to be adapted to Enum
+    # when this is done, there is no need for this requirement
+    @staticmethod
+    def is_proper_direction(vector: Vector) -> bool:
+        direction_vectors = [Direction.NORTH.value, Direction.SOUTH.value, Direction.EAST.value, Direction.WEST.value]
+        return vector in direction_vectors
+
+    @staticmethod
+    def has_proper_tags(scene: 'Scene', object_id: int, tag_names: List[str]) -> bool:
```
Comment (filip-szweda) [inline lynx/common/actions/common_requirements.py]: `has_proper_tags` -> `has_given_tags`
`tag_names` -> `given_tags`
Tags passed as arguments are not always the proper ones

File: lynx/common/actions/common_requirements.py [hunk 5]
```diff
@@ -2,45 +2,57 @@
...
+        object: Object = scene.get_object_by_id(object_id)
+        if len(tag_names) > len(object.tags):
```
Comment (filip-szweda) [inline lynx/common/actions/common_requirements.py]: `return all(element in object.tags for element in tag_names)`
No need for other returns

File: lynx/common/object.py [hunk 6]
```diff
@@ -18,9 +18,8 @@ class Object(Entity):
     position: Vector = Vector(0, 0)
     additional_positions: List[Vector] = field(default_factory=list) 
     state: str = ""
-    walkable: bool = False
     tick: str = ""
     on_death: str = ""
     owner: str = ""
-    pickable: bool = False
-    pushable: bool = False
+    inventory: List[int] = field(default_factory=list)
```
Comment (filip-szweda) [inline lynx/common/object.py]: I think inventory should be implemented in separate pull request, if not Jira issue

File: lynx/common/object.py [hunk 7]
```diff
@@ -18,9 +18,8 @@ class Object(Entity):
...
+    tags: List[str] = field(default_factory=list)       # List of options/extensions, so we don't need to store pickable/pushable in a standalone variable
```
Comment (filip-szweda) [inline lynx/common/object.py]: No need for this comment

File: lynx/common/square.py [hunk 8]
```diff
@@ -12,21 +12,27 @@ class Square:
...
     def append(self, object: Object) -> NoReturn:
         # We might rename the walkable property into sth like ground
-        if object.walkable and self.ground != None:
+        if 'walkable' in object.tags and self.ground != None:
             # In the future when we add logger we should log this event!
             raise Exception("Cannot put more than a one ground in a square!")
-        
-        if object.walkable:
+
+        if 'walkable' in object.tags:
```
Comment (filip-szweda) [inline lynx/common/square.py]: `<tag(s)> in object.tags` is used a lot
I think that we really need `has_tags` method for `Object`

File: tests/common_requirements_test.py [hunk 9]
```diff
@@ -55,3 +55,24 @@ def test_south_vector_proper(self) -> NoReturn:
...
     def test_huge_vector_improper(self) -> NoReturn:
         assert not CommonRequirements.is_proper_direction(Vector(100, 100))
+
+
+class TestContain_tags:
+    scene = Scene()
+    dummy_object1 = Object(id=123, name="dummy", position=Vector(0, 0), tags=['walkable', 'pushable'])
+    scene.add_entity(dummy_object1)
+
+    def test_success(self) -> NoReturn:
```
Comment (filip-szweda) [inline tests/common_requirements_test.py]: Use `test_<test name>_<test circumstances>_<test result>` convention please

**Review by filip-szweda — CHANGES_REQUESTED** [review]
Needs some work, but great thought overall and I'm happy with the logic

File: lynx/common/actions/chop.py (hunk 2 above)
Comment (EasyJB) [reply lynx/common/actions/chop.py]: Didn't we want to walk on logs?

File: lynx/common/actions/common_requirements.py (hunk 5 above)
Comment (EasyJB) [reply lynx/common/actions/common_requirements.py]: Yea, u are right. As I was testing I got some issues, but now it seems that `all() ` is enough.

**Review by blazej-smorawski — APPROVED** [review]

File: lynx/common/actions/chop.py [hunk 10]
```diff
@@ -22,20 +22,19 @@ def apply(self, scene: 'Scene') -> None:
         target_position = self._get_target_position(scene)
         objects_on_square = scene.get_objects_by_position(target_position)
         # TODO put 'Tree' into an Enum
-        chopped_object: Object = \
-        list(filter(lambda object_on_square: object_on_square.name == 'Tree', objects_on_square))[0]
+        chopped_object: Object = list(filter(lambda object_on_square: object_on_square.name == 'Tree',
+                                             objects_on_square))[0]
```
Comment (Krystian030) [inline lynx/common/actions/chop.py]: There we can get index exception, so maybe we should do it like that:
```
chopped_object_iterator = iterator(list(filter(lambda object_on_square: object_on_square.name == 'Tree', objects_on_square))
chopped_object = next(chopped_object_iterator , None)
```

**Review by Krystian030 — CHANGES_REQUESTED** [review]
It is important to adapt the frontend to the changes, so I recommend checking it

File: lynx/common/actions/chop.py (hunk 2 above)
Comment (filip-szweda) [reply lynx/common/actions/chop.py]: Something might have changed, maybe

File: tests/common_requirements_test.py [hunk 11]
```diff
@@ -55,3 +55,24 @@ def test_south_vector_proper(self) -> NoReturn:
...
     def test_huge_vector_improper(self) -> NoReturn:
         assert not CommonRequirements.is_proper_direction(Vector(100, 100))
+
+
+class TestContainTags:
+    scene = Scene()
+    dummy_object1 = Object(id=123, name="dummy", position=Vector(0, 0), tags=['walkable', 'pushable'])
+    scene.add_entity(dummy_object1)
+
+    def test_single_success(self) -> NoReturn:
```
Comment (filip-szweda) [inline tests/common_requirements_test.py]: `test_has_proper_tag_positive`
...
`test_has_improper_tag_negative`
...
`test_has_proper_tags_positive`
...
`test_has_improper_tags_negative`
...
<no need for the last test, you can remove it>

Also, add a test for what happens when you provide an invalid tag e.g.
```
def test_has_invalid_tag_negative(self) -> NoReturn:
   assert not CommonRequirements.has_given_tags(self.scene, 123, ['this-tag-does-not-exist'])
```

Comment (filip-szweda) [reply tests/common_requirements_test.py]: Bumping, I think this and @Krystian030's requested change are the last ones before approval ;)

**Review by Krystian030 — APPROVED** [review]

**Comment by Krystian030** [comment]
@filip-szweda chce to zmergować już, mógłbyś dać approve - żeby Mati mógł później od razu dwie zmiany wrzucić

**Review by filip-szweda — APPROVED** [review]
